### PR TITLE
HTMLImports cleanup

### LINKED ISSUE #3: [v1] `.import` is not an instance of Document
In polyfilled browsers, `link.import instanceof Document` is false

```html
    <script>
        function importLoaded(event) {
          chai.assert.ok(event.target.import instanceof Document, 'load event target can be used to find imported document');
        }
    </script>
    <link rel="import" href="imports/load-smth.html" onload="importLoaded(event)">
```

As per spec https://www.w3.org/TR/html-imports/#interface-import

> the attribute **must** return the _imported document_ for the _import_

and for "_imported document_"

> an _import_ is represented as a _`Document`_, called the **imported document**.

### LINKED ISSUE #7: Execute imports earlier in IE
The current version of HTMLImport polyfill that is shipped in WebComponents polyfill v0.7.x has an issue in IE - html imports are executed with a huge delay. 

Here is detailed discussion: https://github.com/webcomponents/webcomponentsjs/issues/597

In short - the reason is that only for IE polyfill waits when `document.readyState` is `complete`, but not when it is `interactive`.

Are there any plans to fix it? Do you need any help with that?

### LINKED ISSUE #8: Use an anchor instead of new URL
Then we can nuke the `URL` polyfill and live happily ever after.

/cc @sorvell who can help with implementing hints

### LINKED ISSUE #9: HTML Imports polyfill incompatible with Safari's native v1 custom elements
_From @sorvell on December 10, 2016 1:35_

The polyfill mirrors Chrome's native implementation by placing imported dom into disconnected documents. However, v1 custom elements do not customize in these documents. Chrome makes a special exception for HTML Imports and customizes custom elements in them. Therefore, the current approach the polyfill takes results ing custom elements not customizing in imported documents.

To fix this, the polyfill could become less correct by avoiding the use of documents. Instead it could inject hidden dom in the main document. This is the same approach that `polymer build` takes when bundling HTML Imports.

_Copied from original issue: webcomponents/webcomponentsjs#656_

### LINKED ISSUE #11: ensure style reparenting done for IE doesn't break ShadyCSS
#1 will move all imported `<style>` and `<link rel=stylesheet>` to `<head>` in case a `<link rel=stylesheet>` is imported in order to ensure the right cascading order (needed only in IE).

Ensure that this doesn't break ShadyCSS, e.g.
```html
<!-- my-element.html -->
<link rel="stylesheet" href="link.css">
<custom-style>
  <style is="custom-style">
    --my-element-color: red;
  </style>
</custom-style>
```

### PR DESCRIPTION
**Pull request by valdrinkoshi**
Fixes #9, see [README.md](https://github.com/webcomponents/html-imports/pull/1/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8) for intro & rationale. 

## Implementation details

- `Path` handles url resolution
- `Xhr` loads resources and provides a Promise that is resolved when done
- `Loader` has been deleted, deduping is done by `Importer`

### Importer
`Importer` is created only in the polyfilled scenario. It is responsible of monitoring mutations to see if new imports are added. In native html-imports we use a capturing load/error listener to tag imports as loaded.

First, it loads the resources needed by imports. Then, it flattens the imports and dispatches load/error events. Loads are deduped, and flattening is done only when all resources are done loading (see `Importer.load`).

Instead of creating a disconnected `Document` to host imported documents, we append them to the main document in order to leverage the native implementation of Custom Elements, which
expects scripts upgrading the `CustomElementRegistry` to be connected to the
main document.

The host element is the importing link itself. We generate a `DocumentFragment` after the resource is loaded: at this moment we'll also resolve the urls within its children, and add load/error event listeners to styles and scripts.

Scripts won't run if generated through `innerHTML`, so we'll have to clone them after appending them to the main document.

Once all imports are loaded (`Importer.onLoadedAll`), we can:
- add the imports to the main document (see `Importer.flatten`), 
- wait styles loading (see `Importer.waitForStyles`), 
- sequentially run the scripts (see `Importer.runScripts`), 
- finally fire the load/error events for all imports (see `Importer.fireEvents`). 

Note: IE11/Edge requires `<link rel=stylesheet>` to be added as direct children of `<head>` otherwise it will not respect the cascading style order (see https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/10472273/); we disable it by setting type attribute, reparent it, then enable it by removing the type attribute. 

### currentScript

Because all imports are added in the main document, we can rely on the native `currentScript`, and  polyfill it on browsers that don't support it. `_currentScript` has been removed.

## Other fixes

Fixes #11 by providing a placeholder with a pointer to the applied element (see [Caveats / Limitations](https://github.com/webcomponents/html-imports/pull/1/files?short_path=04c6e90#diff-04c6e90faac2675aa89e2176d2eec7d8)).

Fixed #8 by lazily detecting if we can use URL, and if not, use a disconnected document and anchor to retrieve the resolved url.

Fixed #7 by checking only for `document.readyState !== 'loading'`; current tests are passing, but would be good to have more specific ones (separate PR?).

Fixes #3 by handling `<base>` in imports and providing the `importForElement()` method.

## Comparison with v0 implementation

https://github.com/PolymerLabs/html-imports-benchmark

Averages of 30 runs on different browsers

Browser      | v0      | v1      | v1 (es6) | v0-v1 (ms) | v0-v1 (%)     | v1-v1 (es6) (ms) | v1-v1 (es6) (%)
------------ | ------- | ------- | -------- | ---------- | ------------- | ---------------- | ---------------
Chrome 55    | 284.71  | 272.97  | 267.95   | 11.74      | **4.12**      | 5.02             | 1.84
Canary 58    | 252.97  | 238.60  | 227.68   | 14.38      | **5.68**      | 10.92            | 4.58
Firefox 51   | 1026.59 | 742.14  | 690.93   | 284.44     | **27.71**     | 51.22            | 6.90
Safari 10    | 311.11  | 245.68  | 243.22   | 65.43      | **21.03**     | 2.46             | 1.00
Safari TP    | 316.77  | 257.15  | 256.66   | 59.62      | **18.82**     | 0.49             | 0.19
Win10 MSEdge | 1208.06 | 954.23  | 990.17   | 253.83     | **21.01**     | -35.94           | -3.77
Win8.1 IE11  | 1218.34 | 1191.95 | -        | 26.39      | **2.17**      | -                | -

`html-imports.min.js` is 3.29KB gzipped (8.5KB uncompressed)

### CODE REVIEW

File: tests/html/HTMLImports.html [hunk 1]
```diff
@@ -1,51 +0,0 @@
-<!doctype html>
```
Comment (valdrinkoshi) [inline tests/html/HTMLImports.html:1]: Removed because duplicate of dynamic.html tests, and it relies on internal properties.

File: tests/html/customevent-detail.html [hunk 2]
```diff
@@ -1,31 +0,0 @@
-<!doctype html>
```
Comment (valdrinkoshi) [inline tests/html/customevent-detail.html:1]: Removed as it tests the polyfill for CustomEvent, which is not part of html-imports anymore

**Comment by valdrinkoshi** [comment]
@sorvell @justinfagnani @usergenic @cdata FYI, I'll soon update the README with more details.
At this point, all tests are passing, and the only "hack" done here is to workaround IE/Edge styles issues (see description).

File: bower.json [hunk 3]
```diff
@@ -15,8 +15,12 @@
...
   "ignore": [],
+  "dependencies": {
+    "es6-promise": "stefanpenner/es6-promise#^4.0.0",
+    "template": "webcomponents/template#master",
+    "webcomponents-platform": "webcomponents/webcomponents-platform#master"
+  },
   "devDependencies": {
-    "web-component-tester": "^4.0.1",
-    "polymer": "2.0-preview"
+    "web-component-tester": "6.0.0-prerelease.3"
```
Comment (valdrinkoshi) [inline bower.json:23]: forcing this version because of https://github.com/Polymer/web-component-tester/issues/476

File: src/html-imports.js [hunk 4]
```diff
@@ -9,718 +9,754 @@
...
+      const imports = doc.querySelectorAll(IMPORT_SELECTOR);
+      for (let i = 0, l = imports.length; i < l; i++) {
+        whenElementLoaded(imports[i]);
+      }
+      // Observe only document head
+      new MutationObserver(this._onMutation.bind(this)).observe(doc.head, {
+        childList: true
+      });
+
+      if (!useNative) {
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Ideally, if `useNative: true`, as little code is evaluated & run as possible.  AFAIK the only reason the entire polyfill isn't guarded on `useNative` is the `HTMLImports.whenReady()` implementation.

Although the v0 polyfill did provide a "native" implementation of `HTMLImports.whenReady()`, after discussing with @sorvell we think it's probably best that the `html-imports` polyfill doesn't create a `window.HTMLImports` symbol at all.  The "easy-mode" `webcomponents-lite.js` can still create a `HTMLImports.whenReady` stub on native that does a low-fidelity implementation so that users can write browser-agnostic main-document scripts (`<script>HTMLImports.whenReady(() => { ... })</script>`), but in the case only the polyfill is loaded, they would need to detect that the polyfill is in use (`if (window.HTMLImports) ...`) before using the API.  This seems cleaner and more in-line with the other polyfills, and avoids any undue tax on native.

File: src/html-imports.js [hunk 5]
```diff
@@ -9,718 +9,754 @@
...
+          Path.__workingURL = (u.href === 'http://a/c%20d');
+        } catch (e) {}
+      }
+
+      if (Path.__workingURL) {
+        return (new URL(url, base)).href;
+      }
+
+      // Fallback to creating an anchor into a disconnected document.
+      let doc = Path.__tempDoc;
```
Comment (sorvell) [inline src/html-imports.js]: I'm not sure it's worth the code to use the real URL here.

File: src/html-imports.js [hunk 6]
```diff
@@ -9,718 +9,754 @@
...
+      });
+      request.send();
+      return request;
+    }
+  };
+
+  /********************* loader *********************/
+  // This loader supports a dynamic list of urls
+  // and an oncomplete callback that is called when the loader is done.
+  // NOTE: The polyfill currently does *not* need this dynamism or the
```
Comment (sorvell) [inline src/html-imports.js]: Is this comment still true?

File: src/html-imports.js (hunk 4 above)
Comment (sorvell) [inline src/html-imports.js]: Hopefully we don't make an instance of the Importer when using native imports.

Instead, I think we can ask users to guard code on `HTMLImports` existing and define that global only when the polyfill is in use.

File: src/html-imports.js [hunk 7]
```diff
@@ -9,718 +9,754 @@
...
+      flags.log && console.log('loaded', url, elt);
+      // We've already seen a document at this url, return.
+      if (this.documents[url] !== undefined) {
+        return;
+      }
+      if (err) {
+        this.documents[url] = null;
+      } else {
+        // Generate an HTMLDocument from data.
+        const doc = makeDocument(resource, redirectedUrl || url);
```
Comment (sorvell) [inline src/html-imports.js]: Why isn't `makeDocument` a member of the `Importer` class?

File: src/html-imports.js [hunk 8]
```diff
@@ -9,718 +9,754 @@
...
-      }
+
+  /**
+   * Creates a new document containing resource and normalizes urls accordingly.
+   * @param {string=} resource
+   * @param {string=} url
+   * @return {HTMLElement}
+   */
+  function makeDocument(resource, url) {
+    const content = /** @type {HTMLElement} */
```
Comment (sorvell) [inline src/html-imports.js]: Let's consider making this a `documentFragment`. The main reason it was an element was to hide it easily, but now that it's a child of the link element, the content will be auto-hidden.

Then we'd just set `import.link = import`

File: src/html-imports.js [hunk 9]
```diff
@@ -9,718 +9,754 @@
...
-var scriptType = "import-pending";
+  /**
+   * Clones styles and stylesheets links contained in imports and moves them
+   * as siblings of the root import link.
+   * @param {!HTMLLinkElement} importLink
+   */
+  function cloneAndMoveStyles(importLink) {
+    const n$ = importLink.import.querySelectorAll(stylesSelector);
+    for (let i = 0, l = n$.length, n; i < l && (n = n$[i]); i++) {
+      const clone = document.createElement(n.localName);
```
Comment (sorvell) [inline src/html-imports.js]: Why is the element being cloned? 

And if the clone ie needed, why not use `cloneNode(true)`?

File: src/html-imports.js [hunk 10]
```diff
@@ -9,718 +9,754 @@
...
+
+    /**
+     * @param {Array<MutationRecord>} mutations
+     */
+    _onMutation(mutations) {
+      for (let j = 0, m; j < mutations.length && (m = mutations[j]); j++) {
+        for (let i = 0, l = m.addedNodes ? m.addedNodes.length : 0; i < l; i++) {
+          const n = /** @type {Element} */ (m.addedNodes[i]);
+          if (n && isImportLink(n)) {
+            if (useNative) {
```
Comment (sorvell) [inline src/html-imports.js]: Again, hopefully none of this code is running if native imports are available.

File: src/html-imports.js [hunk 11]
```diff
@@ -9,718 +9,754 @@
...
-              }
+
+    /**
+     * @param {Array<MutationRecord>} mutations
+     */
+    _onMutation(mutations) {
+      for (let j = 0, m; j < mutations.length && (m = mutations[j]); j++) {
+        for (let i = 0, l = m.addedNodes ? m.addedNodes.length : 0; i < l; i++) {
+          const n = /** @type {Element} */ (m.addedNodes[i]);
+          if (n && isImportLink(n)) {
```
Comment (sorvell) [inline src/html-imports.js]: I guess nothing needs to be done if a script or style is created dynamically inside an import, but do we have a test for that? (and was this supported in the old polyfill?)

File: src/html-imports.js [hunk 12]
```diff
@@ -9,718 +9,754 @@
...
-          n.appendChild(n.import);
-          if (document.contains(n.parentNode)) {
-            // TODO(sorvell): need to coordinate with observer in document.head.
-            //this.observe(n.import);
+    }
+
+    _onLoadedAll() {
+      this._flatten(this._doc);
+      Promise.all([
+        runScripts(this._doc),
```
Comment (sorvell) [inline src/html-imports.js]: There is no ordering here between scripts and styles which is technically incorrect. We can probably get away with this but it should be documented at the least, is it?

File: src/html-imports.js [hunk 13]
```diff
@@ -9,718 +9,754 @@
...
+        waitForStyles(this._doc)
+      ]).then(() => fireEvents(this._doc));
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Just noting, since `runScripts` and `waitForStyles` will execute in parallel, this loses a guarantee that styles are applied before scripts run.  @sorvell says this probably isn't required fidelity for the polyfill (since computing style is unreliable during document loading).  Maybe just worth a comment in case this comes up as an issue?

File: src/html-imports.js [hunk 14]
```diff
@@ -9,718 +9,754 @@
...
+   */
+  function isImportLink(node) {
+    return node.nodeType === Node.ELEMENT_NODE && MATCHES.call(node, IMPORT_SELECTOR);
+  }
 
-// loader singleton to handle loading imports
-var importLoader = new Loader(importer.loaded.bind(importer),
-    importer.loadedAll.bind(importer));
+  /********************* vulcanize style inline processing  *********************/
+  const attrs = ['action', 'src', 'href', 'url', 'style'];
```
Comment (sorvell) [inline src/html-imports.js]: It's not clear what the difference is between these url related functions and the `Path` object above. Might they be merged together?

File: src/html-imports.js [hunk 15]
```diff
@@ -9,718 +9,754 @@
...
-  var content = template.content;
-  var n$ = content.querySelectorAll('style, form[action], [src], [href], [url], [style]');
-  for (var i=0; i < n$.length; i++) {
-    var n = n$[i];
-    if (n.localName == 'style') {
-      path.resolveUrlsInStyle(n, base);
-    } else {
-      fixUrlAttributes(n, base);
+
+  function fixDomModules(element, url) {
```
Comment (sorvell) [inline src/html-imports.js]: Needs a comment about this being Polymer specific.

File: src/html-imports.js [hunk 16]
```diff
@@ -9,718 +9,754 @@
...
+   * Updates the `currentScript`.
+   * @param {!HTMLDocument} doc
+   * @return {Promise} Resolved when scripts are loaded.
+   */
+  function runScripts(doc) {
+    const s$ = doc.querySelectorAll(`script[type=${scriptType}]`);
+    let promise = Promise.resolve();
+    for (let i = 0, l = s$.length, s; i < l && (s = s$[i]); i++) {
+      promise = promise.then(() => {
+        const c = doc.createElement('script');
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Let's add a note here to indicate `s.coneNode(true)` is insufficient as it was originally created via `innerHTML`, and those scripts can't be cloned and run.

File: src/html-imports.js [hunk 17]
```diff
@@ -9,718 +9,754 @@
...
   }
-  fixUrlsInTemplates(content, base);
-}
 
-function fixUrlsInTemplates(element, base) {
-  var t$ = element.querySelectorAll('template');
-  for (var i=0; i < t$.length; i++) {
-    fixUrlsInTemplate(t$[i], base);
+  /**
+   * Replaces all the imported scripts with a clone in order to execute them.
```
Comment (sorvell) [inline src/html-imports.js]: Add comment about why this is needed (for security, this is the only way to run a script that has been innerHTML'd...)

Comment (sorvell) [reply src/html-imports.js]: Let's make sure there is a test that ensures that an import that is dynamically inserted into another one does not cause scripts to double execute.

File: src/html-imports.js [hunk 18]
```diff
@@ -9,718 +9,754 @@
...
-        urlPath = (new URL(urlPath, linkUrl)).href;
-      }
-      return pre + '\'' + urlPath + '\'' + post;
+  // Polyfill `currentScript` for browsers without it.
+  let currentScript = null;
+  if ('currentScript' in document === false) {
+    Object.defineProperty(document, 'currentScript', {
+      get: function() {
+        return currentScript ||
+          // NOTE: only works when called in synchronously executing code.
```
Comment (sorvell) [inline src/html-imports.js]: For clarity, let's add a note here that the `||` doesn't have to do with HTMLImports but helps us better generally polyfill `currentScript`.

File: src/html-imports.js [hunk 19]
```diff
@@ -9,718 +9,754 @@
...
+        }
+        // Listen for load/error events before adding the clone to the document.
+        // Catch failures, always return c.
+        const whenLoadedPromise = whenElementLoaded(c).catch(() => c);
+        // Update currentScript and replace original with clone script.
+        currentScript = c;
+        s.parentNode.replaceChild(c, s);
+        // After is loaded, reset currentScript.
+        return whenLoadedPromise.then((script) => {
+          if (script === currentScript) {
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Do you ever expect this to be false?  If so, seems like it would indicate an error condition?  AFAICT the scripts are run sequentially and this is the only place where `currentScript` is reset...

File: src/html-imports.js [hunk 20]
```diff
@@ -9,718 +9,754 @@
...
-  fixUrlsInTemplates(element, base);
-}
 
-var scriptType = "import-pending";
+  /**
+   * Clones styles and stylesheets links contained in imports and moves them
+   * as siblings of the root import link.
+   * @param {!HTMLLinkElement} importLink
+   */
+  function cloneAndMoveStyles(importLink) {
```
Comment (kevinpschaaf) [inline src/html-imports.js]: In the README, let's add a "Caveats / Limitations" section that describes this particular caveat (style elements may be moved out of the import document on IE/Edge, thus changing the import document's structure-- e.g. `document.currentScript.previousElement` may now be different).  In that section, you could also consolidate the other caveats into a list, i.e. that the `<link>.import` is not actually a document, dynamic imports must be added to head, etc.

File: src/html-imports.js [hunk 21]
```diff
@@ -9,718 +9,754 @@
...
+        const csr = element.sheet.cssRules;
+        // search the rules for @import's
+        for (let i = 0, l = csr ? csr.length : 0; i < l && isLoaded; i++) {
+          if (csr[i].type === CSSRule.IMPORT_RULE) {
+            // if every @import has resolved, fake the load
+            isLoaded = Boolean(csr[i].styleSheet);
+          }
+        }
       }
+    } else if (element.localName === 'script' && !element.src) {
```
Comment (sorvell) [inline src/html-imports.js]: Why is this needed? don't scripts always fire a load event... and can inline scripts execute async?

File: src/html-imports.js (hunk 15 above)
Comment (kevinpschaaf) [inline src/html-imports.js]: Let's remove this function to eliminate the Polymer coupling, and instead add a general polyfill API that `dom-module` will call if the `HTMLImports` polyfill is loaded to find the original location of its import, something like:

```
HTMLImports.importForElement = function(el) {
  // walk up to find `<link rel="import">` and return that (or null)
}
```

Then, `dom-module` would need to gain a `resolveURL` API that calls `HTMLImports.importForElement` and returns a resolved URL.

File: src/html-imports.js (hunk 21 above)
Comment (sorvell) [reply src/html-imports.js]: Won't this fail to notice that scripts and styles have loaded? I guess this is not required because of how it's used but this should probably be spelled out.

File: src/html-imports.js [hunk 22]
```diff
@@ -9,718 +9,754 @@
...
-        handleImport(n);
-      }
+    // Support <base> in imported docs. Resolve url and remove it from the parent.
+    const baseEl = /** @type {HTMLBaseElement} */ (content.querySelector('base'));
+    if (baseEl) {
+      url = Path._resolveUrl(baseEl.getAttribute('href'), url);
+      baseEl.parentNode.removeChild(baseEl);
     }
+    // TODO(sorvell): this is specific to users of <dom-module> (Polymer).
+    fixDomModules(content, url);
```
Comment (sorvell) [inline src/html-imports.js]: Let's see if we can remove this. Instead HTMLImports can add a custom api `importForElement`.

File: src/html-imports.js [hunk 23]
```diff
@@ -9,718 +9,766 @@
...
             }
           }
-        });
-      });
-      element.__importObserver.observe(element, {childList: true, subtree: true});
+        }
+      }
+      if (promises.length) {
+        // Ensure we update isImporting.
+        isImporting = true;
```
Comment (sorvell) [inline src/html-imports.js]: This technique will not work when `whenReady` is used before the mutation observer notices the import. For example:

```
document.head.appendChild(link);
HTMLImports.whenReady(callback):
```

File: src/html-imports.js [hunk 24]
```diff
@@ -9,718 +9,754 @@
...
+        c.textContent = s.textContent;
+        if (s.src) {
+          c.setAttribute('src', s.getAttribute('src'));
+        }
+        // Listen for load/error events before adding the clone to the document.
+        // Catch failures, always return c.
+        const whenLoadedPromise = whenElementLoaded(c).catch(() => c);
+        // Update currentScript and replace original with clone script.
+        currentScript = c;
+        s.parentNode.replaceChild(c, s);
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Let's add a comment somewhere (either here or in `whenElementLoaded`) that you're relying on `replaceChild` of an inline script to run the script synchronously before returning, since `whenElementLoaded` doesn't wait for the load event since there is no `src`.

File: src/html-imports.js (hunk 23 above)
Comment (sorvell) [reply src/html-imports.js]: This also doesn't catch imports added to the main document does it?

**Review by sorvell — CHANGES_REQUESTED** [review]

File: src/html-imports.js (hunk 9 above)
Comment (kevinpschaaf) [inline src/html-imports.js]: Is there a reason `cloneNode` won't work for cloning the style?  If so, add comment?

**Review by kevinpschaaf — CHANGES_REQUESTED** [review]

File: src/html-imports.js (hunk 5 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Did some tests to compare creating 1000 URLs vs creating 1 doc and reusing it for urls, and url method (even when polyfilled) is ~1.5-2x more efficient. Tested with https://gist.github.com/valdrinkoshi/195c1dfc44ffc164b96c0d6cc1d65530

File: src/html-imports.js (hunk 6 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: That's part of the potential improvement that could be done (in a separate PR). I've given it a shot to make it all promises-based but stumbled upon deduping issues, so I've postponed that work.

File: src/html-imports.js (hunk 12 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: I should probably first wait for the styles, then run the scripts. Will update this!

File: src/html-imports.js (hunk 7 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Refactoring TBD!

File: src/html-imports.js (hunk 14 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: 👌  will do

File: src/html-imports.js (hunk 17 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: That shouldn't happen as we replace the `type="import-script"` scripts, but yeah I'll add a test!

File: src/html-imports.js (hunk 19 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: I don't think this is needed, was added while testing IE11. Will remove

File: src/html-imports.js (hunk 21 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Remainings of IE11 testing...should be removed and/or documented

File: src/html-imports.js (hunk 18 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: 👌  will do

File: src/html-imports.js (hunk 13 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: I believe it's safer to wait for the styles to load then execute the scripts, and finally fire the events. Will make it slower but more correct

File: src/html-imports.js (hunk 11 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: `v0` handled only imports (see https://github.com/webcomponents/html-imports/blob/v0/src/dynamic.js).

File: src/html-imports.js (hunk 10 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: yep, will remove

File: src/html-imports.js (hunk 16 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: sg

File: src/html-imports.js (hunk 24 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: I should check again for scripts w/o src, if we can rely on them firing load/error.

File: src/html-imports.js (hunk 9 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: sg, will document

File: src/html-imports.js (hunk 8 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: oh right, will try

File: src/html-imports.js (hunk 9 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Clones of links that already had a parent won't work, will document https://gist.github.com/valdrinkoshi/4a92f97169a6fc41a1852f23211b8c4e

File: src/html-imports.js (hunk 6 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: will remove this comment as the polyfill currently relies on this. The improvement i had in mind might not be possible/easy as we need to do deduping.

**Comment by valdrinkoshi** [comment]
@kevinpschaaf @sorvell incorporated most of the feedback, PTAL

File: README.md [hunk 25]
```diff
@@ -1 +1,90 @@
...
+
+This is how the resolved import will look like:
+
+```html
+<head>
+  <link rel="stylesheet" href="my-linked-style.css">
+  <style> .blue { color: blue }; </style>
+  <link rel="import" href="my-element.html">
+    <link>
+    <style></style>
```
Comment (kevinpschaaf) [inline README.md]: Let's add a `type="import-placeholder"` or something similar, so that they don't show up in `document.styleSheets` (adding a `type` makes them not parse).

File: src/html-imports.js (hunk 18 above)
Comment (kevinpschaaf) [reply src/html-imports.js]: Looks good.

File: src/html-imports.js [hunk 26]
```diff
@@ -9,718 +9,780 @@
...
+          Path.replaceAttrUrl(element.getAttribute('href'), base));
+      }
+      if (element.src) {
+        element.setAttribute('src',
+          Path.replaceAttrUrl(element.getAttribute('src'), base));
+      }
+      if (element.localName === 'style') {
+        Path.resolveUrlsInStyle(element, base);
+      } else if (element.localName === 'script' && element.textContent) {
+        element.textContent += `\n//# sourceURL=${base}`;
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Note (non-blocking): Let's test the "debuggability" of scripts loaded via imports on the various browsers:
* Can you set breakpoints?
* Can you find a script by name?
* How are syntax errors surfaced in the debugger?

Traditionally it's been hard to have all of these, iirc.  The previous polyfill used data-uri's which were good on some aspects and bad on others (forget the exact details).

File: src/html-imports.js [hunk 27]
```diff
@@ -9,718 +9,780 @@
...
+      }
+    },
+
+    fixUrlAttributes: function(element, base) {
+      const attrs = ['action', 'src', 'href', 'url', 'style'];
+      for (let i = 0, a; i < attrs.length && (a = attrs[i]); i++) {
+        const at = element.attributes[a];
+        const v = at && at.value;
+        if (v && (v.search(/({{|\[\[)/) < 0)) {
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Please add comment: This is a wart (because it's binding syntax-specific) that really shouldn't be here; no great ideas on how to solve this... a library-implementable hook is one option but there wouldn't be a convenient time for the library to implement it.  So... let's just add a comment and not get stuck on it.

File: src/html-imports.js (hunk 5 above)
Comment (kevinpschaaf) [reply src/html-imports.js]: SGTM

File: src/html-imports.js [hunk 28]
```diff
@@ -9,718 +9,780 @@
...
+          }
+          // Replace the element we're about to move with a placeholder
+          // containing a reference to the moved element.
+          const placeholder = document.createElement(s.localName);
+          placeholder['__appliedElement'] = s;
+          s.parentNode.replaceChild(placeholder, s);
+          if (rootImport.parentNode === document.head) {
+            document.head.insertBefore(s, rootImport);
+          } else {
+            document.head.appendChild(s);
```
Comment (kevinpschaaf) [inline src/html-imports.js]: Can these not go in `<body>`?

Moving these to head has the potential to break cascade order:
```html
<html>
  <head></head>
  <body>
    <style>/*...*/</style>
    <link rel="import" href="styles-will-be-moved-into-head.html">
  </body>
</html>
```

File: src/html-imports.js [hunk 29]
```diff
@@ -9,718 +9,780 @@
...
-  function bootstrap() {
-    importer.bootDocument(document);
+  /**
+   * Returns the link that imported the element.
+   * @param {!Element} element
+   * @return {!HTMLLinkElement|undefined}
+   */
+  function importForElement(element) {
+    let target = element;
+    while ((target = target['__ownerImport'] || target.parentNode || target.host)) {
```
Comment (kevinpschaaf) [inline src/html-imports.js]: The memoization here could be slightly cleaner.  Right now, even once you've memoized `__ownerImport`, you still go into the while loop and check if it's an import link.  Would be slightly better to fast-path if already memoized.  Pretty minor.

**Review by kevinpschaaf — CHANGES_REQUESTED** [review]
Let's do a quick review of these latest ones and prioritize... most are minor.

File: README.md (hunk 25 above)
Comment (valdrinkoshi) [reply README.md]: good catch, done 👌

File: src/html-imports.js (hunk 26 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Commented this for now and added a TODO for future valdrin

File: src/html-imports.js (hunk 27 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Done

File: src/html-imports.js (hunk 29 above)
Comment (valdrinkoshi) [reply src/html-imports.js]: Yeah it was a bit cryptic, fixed!

**Comment by valdrinkoshi** [comment]
safari tests seem flaky, apparently safari downloads and sets stylesheets even if they have a `type` attribute 😭 
```html
<head>
  <link rel=stylesheet type="disabled" href="import.css" 
        onload="console.log(event)" onerror="console.log(event)">
</head>

File: src/html-imports.js [hunk 30]
```diff
@@ -9,718 +9,742 @@
...
+   */
+  class Importer {
+    constructor() {
+      this.documents = {};
+      // Used to keep track of pending loads, so that flattening and firing of
+      // events can be done when all resources are ready.
+      this.inflight = 0;
+      whenDocumentReady(() => {
+        // Observe only document head
+        new MutationObserver(this._onMutation.bind(this)).observe(document.head, {
```
Comment (valdrinkoshi) [inline src/html-imports.js]: Do this at constructor time to catch mutations as early as possible

File: src/html-imports.js [hunk 31]
```diff
@@ -9,718 +9,742 @@
...
-    if (o.src) {
-      var src = path.replaceAttrUrl(o.getAttribute('src'), o.__baseURI);
-      c.setAttribute('src', src);
+    /**
+     * Fires load/error event for the import if this wasn't done already.
+     * @param {!HTMLLinkElement} link
+     */
+    _fireEventIfNeeded(link) {
+      // Don't fire twice same event.
+      if (!link['__loaded']) {
```
Comment (valdrinkoshi) [inline src/html-imports.js]: Set the readyState to complete

File: src/html-imports.js [hunk 32]
```diff
@@ -9,718 +9,742 @@
...
+      const n$ = /** @type {!NodeList<!HTMLLinkElement>} */
+        (doc.querySelectorAll(importSelector));
+      for (let i = 0, l = n$.length, n; i < l && (n = n$[i]); i++) {
+        const imp = this.documents[n.href];
+        n.import = /** @type {!Document} */ (imp);
+        if (imp && imp.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+          this.documents[n.href] = n;
+          // Suppress Closure warning about incompatible subtype assignment.
+          ( /** @type {!HTMLElement} */ (n).import = n);
+          this._flatten(imp);
```
Comment (valdrinkoshi) [inline src/html-imports.js]: Set n.readyState to loading

File: README.md [hunk 33]
```diff
@@ -1 +1,105 @@
...
+    <link type="import-placeholder">
+    <style type="import-placeholder"></style>
+  </link>
+</head>
+```
+
+The placeholders contain a reference to the applied element:
+
+```javascript
+var myImport = document.head.querySelector('link[rel=import]');
```
Comment (valdrinkoshi) [inline README.md]: access to `myImport.import` here

**Review by sorvell — APPROVED** [review]

File: tests/polyfills.js [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+// Includes CustomEvent polyfill.
+document.write('<script src="/bower_components/webcomponents-platform/webcomponents-platform.js"></script>');
```
Comment (sorvell) [inline tests/polyfills.js]: This should use canonical path which is something like `../../` rather than `/bower_components/`

File: src/html-imports.js [hunk 35]
```diff
@@ -7,720 +7,756 @@
...
-                importer.loadNode(p);
-              }
-            }
+
+    /**
+     * Replaces all the imported scripts with a clone in order to execute them.
+     * Updates the `currentScript`.
+     * @return {Promise} Resolved when scripts are loaded.
+     */
+    runScripts() {
```
Comment (sorvell) [inline src/html-imports.js:460]: Re-ping on status of debugging when scripts are run via this method. I believe @kevinpschaaf asked about this.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: src/html-imports.js (hunk 35 above)
Comment (valdrinkoshi) [reply src/html-imports.js:460]: Seems like in FF failing scripts will break the link to the sourceURL...guess i have to set the src as dataurl with the text content instead..!

Comment (valdrinkoshi) [reply src/html-imports.js:460]: Added explainer in `makeDocument` of why we set `src` instead of using `textContent`

File: tests/polyfills.js (hunk 34 above)
Comment (valdrinkoshi) [reply tests/polyfills.js]: Done

**Review by kevinpschaaf — APPROVED** [review]
